viper.go

```diff
@@ -864,22 +864,28 @@ func (v *Viper) GetSizeInBytesE(key string) (uint, error) {
 func UnmarshalKey(key string, rawVal interface{}, opts ...DecoderConfigOption) error {
 	return v.UnmarshalKey(key, rawVal, opts...)
 }
-
 func (v *Viper) UnmarshalKey(key string, rawVal interface{}, opts ...DecoderConfigOption) error {
-	// We first get all value we have for a key (from default to override)
 	lcaseKey := strings.ToLower(key)
-	values := v.findAll(lcaseKey)
 
-	// findAll returns all the value for a settings from highest priority to lowest. So when aggregating them we want
-	// to start at the lowest priority and override with higher ones.
-	for idx := len(values) - 1; idx >= 0; idx-- {
-		err := decode(values[idx], defaultDecoderConfig(rawVal, opts...))
-		if err != nil {
-			return err
+	// AllSettings returns settings from every sources merged into one tree
+	settings := v.AllSettings()
+
+	keyParts := strings.Split(lcaseKey, v.keyDelim)
+	for i := 0; i < len(keyParts)-1; i++ {
+		if setting, found := settings[keyParts[i]]; found {
```

Comment:
Personally, I find it very confusing to have distinct variables with names only that are only 1 letter different. Rename it to make it clear that there's no shadowing going on.

```suggestion
		if value, found := settings[keyParts[i]]; found {
```